src/main/java/com/rtb/projectmanagementtool/task/TaskController.java

```diff
@@ -13,22 +13,34 @@ public TaskController(DatastoreService datastore) {
     this.datastore = datastore;
   }
 
+  public ArrayList<Key> addTasks(ArrayList<TaskData> tasks) {
+    ArrayList<Entity> taskEntities = new ArrayList<>();
+    for (TaskData task : tasks) {
+      taskEntities.add(task.toEntity());
+    }
+    return new ArrayList<>(datastore.put(taskEntities));
+  }
+
+  public ArrayList<Key> addSubtasks(TaskData task, ArrayList<TaskData> subtasks) {
+    ArrayList<Key> keys = addTasks(subtasks);
+    ArrayList<Long> taskSubtasks = task.getSubtasks();
+    for (Key key : keys) {
+      taskSubtasks.add(key.getId());
+    }
+    task.setSubtasks(taskSubtasks);
```

Comment:
We can have every subtask point to its parent task instead (and remove the subtaskID list in the parent task). That would still show the association without updating the parent task. However, that would be really inefficient when trying to retrieve the task's subtasks.

I know ancestor paths can be created in datastore. Is that something I should look into using instead?

Reply:
Yep subtasks with a pointer to the parent task would be a more standard approach.  I agree efficiency might suffer specially if you have multiple levels you need to load (since that'll mean multiple recursive queries), but you do get data robustness in exchange.  I think this should be ok if we're only displaying 1 level on the screen (parent plus its children).

One thing you can do to help with efficiency if we're displaying the whole tree would be to set a "root" ID on the subtasks as well (parent of the parent etc..), which in our case would point to the ProjectID.  You can then load the whole project (single query) and build your tree in Java.  You'd end up pulling tasks you probably don't need (eg sibling tasks), but it would still be better than loading all tasks in the datastore.


Regarding ancestor paths in keys, I read about it a bit as well.  It sounds like it's designed more for locking a family of nodes, and would work well when there's a small number of members in the family.  Groups with a bigger number of members would mean the whole family tree would get locked when loading any of its members.  It could be ok for our scenario if the assumption is a project would not have millions of tasks assigned to it (eg. if we assume a max of 50 tasks per project).